File: adapter-pattern/java/src/PlayerAdapter.java
package src;

// 适配器类，可以播放各种音频和视频，外部调用方式一样。适配器也可以实现基础播放接口
public class PlayerAdapter implements MediaPlayer {

  private String mediaType;
  private String fileType;
  private String filePath;
  private MediaPlayer mp4Player;
  private MediaPlayer hlsPlayer;
  private AudioPlayer audioPlayer;

  public PlayerAdapter(String mediaType, String fileType, String filePath) {
    this.mediaType = mediaType;
    this.fileType = fileType;
    if (mediaType.equals("video")) {
      if (fileType.equals("mp4")) {
        setMp4Player(fileType, filePath);
      } else if (fileType.equals("hls")) {
        setHlsPlayer(fileType, filePath);
      }
    } else if (mediaType.equals("audio")) {
      setAudioPlayer(fileType, filePath);
    }
  }

  public void setMp4Player(String fileType, String filePath) {
    if (mp4Player == null) {
      mp4Player = new Mp4Player(fileType, filePath);
    } else {
      mp4Player.setFileType(fileType);
      mp4Player.setFilePath(filePath);
    }
  }

  public void setHlsPlayer(String fileType, String filePath) {
    if (hlsPlayer == null) {
      hlsPlayer = new HlsPlayer(fileType, filePath);
    } else {
      hlsPlayer.setFileType(fileType);
      hlsPlayer.setFilePath(filePath);
    }
  }

  public void setAudioPlayer(String fileType, String filePath) {
    if (audioPlayer == null) {
      audioPlayer = new AudioPlayer(fileType, filePath);
    } else {
      audioPlayer.setFileType(fileType);
      audioPlayer.setFilePath(filePath);
    }
  }

  // 通用的适配接口
  public void play() {
    if (this.mediaType.equals("audio")) {
      audioPlayer.play();
    } else if (this.mediaType.equals("video")) {
      if (this.fileType.equals("mp4")) {
        mp4Player.play();
      } else if (this.fileType.equals("hls")) {
        hlsPlayer.play();
      }
    }
  }

  // 根据路径播放，可根据需要复用实例
  public void play(String mediaType, String fileType, String filePath) {
    if (mediaType.equals("audio")) {
      setAudioPlayer(fileType, filePath);
      audioPlayer.play();
    } else if (mediaType.equals("video")) {
      if (fileType.equals("mp4")) {
        setMp4Player(fileType, filePath);
        mp4Player.play();
      } else if (fileType.equals("hls")) {
        setHlsPlayer(fileType, filePath);
        hlsPlayer.play();
      }
    }
  }

  public void setFileType(String fileType) {
    this.fileType = fileType;
  }

  public void setFilePath(String filePath) {
    this.filePath = filePath;
  }

}

File: adapter-pattern/js/src/PlayerAdapter.js
import { MediaPlayer } from './MediaPlayer.js'
import { HlsPlayer } from './HlsPlayer.js'
import { Mp4Player } from './Mp4Player.js'
import { AudioPlayer } from './AudioPlayer.js'

// 适配器类，可以播放各种音频和视频，外部调用方式一样。适配器继承基类可选。
// export class PlayerAdapter extends MediaPlayer {
export class PlayerAdapter {
  constructor(mediaType, fileType, filePath) {
    // super(fileType, filePath)
    this.mediaType = mediaType
    this.fileType = fileType
    this.mp4Player = null
    this.hlsPlayer = null
    this.audioPlayer = null
    if (mediaType === 'video') {
      if (fileType === 'mp4') {
        this.setMp4Player(fileType, filePath)
      } else if (fileType === 'hls') {
        this.setHlsPlayer(fileType, filePath)
      }
    } else if (mediaType === 'audio') {
      this.setAudioPlayer(fileType, filePath)
    }
  }

  setMp4Player(fileType, filePath) {
    if (!this.mp4Player) {
      this.mp4Player = new Mp4Player(fileType, filePath)
    } else {
      this.mp4Player.setFileType(fileType)
      this.mp4Player.setFilePath(filePath)
    }
  }

  setHlsPlayer(fileType, filePath) {
    if (!this.hlsPlayer) {
      this.hlsPlayer = new HlsPlayer(fileType, filePath)
    } else {
      this.hlsPlayer.setFileType(fileType)
      this.hlsPlayer.setFilePath(filePath)
    }
  }

  setAudioPlayer(fileType, filePath) {
    if (!this.audioPlayer) {
      this.audioPlayer = new AudioPlayer(fileType, filePath)
    } else {
      this.audioPlayer.setFileType(fileType)
      this.audioPlayer.setFilePath(filePath)
    }
  }

  // 通用的适配接口
  play() {
    if (this.mediaType === 'audio') {
      this.audioPlayer.play()
    } else if (this.mediaType === 'video') {
      if (this.fileType === 'mp4') {
        this.mp4Player.play()
      } else if (this.fileType === 'hls') {
        this.hlsPlayer.play()
      }
    }
  }

  // 根据路径播放，可根据需要复用实例
  playBy(mediaType, fileType, filePath) {
    if (mediaType === 'audio') {
      this.setAudioPlayer(fileType, filePath)
      this.audioPlayer.play()
    } else if (mediaType === 'video') {
      if (fileType === 'mp4') {
        this.setMp4Player(fileType, filePath)
        this.mp4Player.play()
      } else if (fileType === 'hls') {
        this.setHlsPlayer(fileType, filePath)
        this.hlsPlayer.play()
      }
    }
  }
}
